src/lib/distances.jl

```diff
@@ -47,10 +47,10 @@ end
   end
 
 @adjoint function (::Euclidean)(x::AbstractVector, y::AbstractVector)
-  D = x.-y
+  D = x .- y
   δ = sqrt(sum(abs2, D))
   function euclidean(Δ::Real)
-    x̄ = (Δ / δ) .* D
+    x̄ = ifelse(iszero(δ), D, (Δ / δ) .* D)
```

Comment:
Isn't this mathematically incorrect if `x !== y`? E.g., in the 1D case the derivative of the Euclidean distance `x -> |x - a|` with respect to `x` is not defined for `x = a` since the limit from the left and right do not agree (they are -1 and 1, so also not close to 0).

I understand that this might be unsatisfying for practical applications, also due to floating point issues, but it seems mathematically actually the NaN value resulting from the previous implementation makes sense if `x !== y`.

Reply:
we tend to use a subgradient-based convention that if gradient limit from the left has opposite sign to the gradient approaching from the right.
Or more generally: if 0 is an element of the subgradient,
then we say the gradient is zero.

This has the benificial effect of ensuring that the gradient is considered to be zero at local minima,
which from perspective of doing gradient based optimisation is what you want.

Reply:
I see :+1: Thanks for the explanation!